### PR TITLE
Avoid string out of bounds error in snapshot delete

### LINKED ISSUE #8771: [Flaky Test] SnapshotResiliencyTests.testConcurrentSnapshotDeleteAndDeleteIndex if flaky
https://build.ci.opensearch.org/job/gradle-check/20332/

```
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.snapshots.SnapshotResiliencyTests.testConcurrentSnapshotDeleteAndDeleteIndex" -Dtests.seed=B7F15DFF786CE73D -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es-NI -Dtests.timezone=Asia/Oral -Druntime.java=17
NOTE: leaving temporary files on disk at: /var/jenkins/workspace/gradle-check/search/server/build/testrun/test/temp/org.opensearch.snapshots.SnapshotResiliencyTests_B7F15DFF786CE73D-001
NOTE: test params are: codec=Asserting(Lucene95): {}, docValues:{}, maxPointsInLeafNode=330, maxMBSortInHeap=6.1702426242178525, sim=Asserting(RandomSimilarity(queryNorm=true): {}), locale=es-NI, timezone=Asia/Oral
NOTE: Linux 5.15.0-1036-aws amd64/Eclipse Adoptium 17.0.7 (64-bit)/cpus=96,threads=1,free=189188400,total=536870912
```

### PR DESCRIPTION
**Pull request by andrross**
Test failure #8771 shows cases where certain random seeds trigger this case. The bug is clear: the substring() call should happen after the startsWith() check in case the blob name is shorter than the prefix length being used as the start index of the substring call. I don't yet know if/how this manifests in real deployments.

This bug is deterministically reproducible with the following seed, which also verifies the fix:

```
./gradlew ':server:test' --tests "org.opensearch.snapshots.SnapshotResiliencyTests.testConcurrentSnapshotDeleteAndDeleteIndex" -Dtests.seed=E816C7242330BF50
```

### Related Issues
Resolves #8771

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by andrross** [comment]
FYI @sohami @harishbhakuni @kasundra07

File: server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -1592,8 +1592,11 @@ private void executeOneStaleIndexDelete(
                                 Map<String, BlobMetadata> shardLevelBlobs = shardBlob.getValue().listBlobs();
                                 for (Map.Entry<String, BlobMetadata> shardLevelBlob : shardLevelBlobs.entrySet()) {
                                     String blob = shardLevelBlob.getKey();
-                                    String snapshotUUID = blob.substring(SHALLOW_SNAPSHOT_PREFIX.length(), blob.length() - ".dat".length());
                                     if (blob.startsWith(SHALLOW_SNAPSHOT_PREFIX) && blob.endsWith(".dat")) {
+                                        String snapshotUUID = blob.substring(
```
Comment (kotwanikunal) [inline server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java:1596]: Looks like this code is duplicated on L#3365. Would it make sense to pull it out into a common function?
I know the logic will not change frequently enough, but a common location will ensure consistency.

Comment (andrross) [reply server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java]: Updated. Made a few other minor/cosmetic changes as well. Note the diff is much easier to follow if you hide whitespace changes as I removed an outer condition that changed the indent.

**Review by kotwanikunal — APPROVED** [review]

**Review by deshsidd — APPROVED** [review]

**Review by kotwanikunal — APPROVED** [review]

**Comment by andrross** [comment]
> ❌ Gradle check result for [09778ad](https://github.com/opensearch-project/OpenSearch/commit/09778ad3e563fffeb6a0d76677fbbf93ec5762c1): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/33824/)

#9208
#9191
#10673

**Review by harishbhakuni — APPROVED** [review]

**Comment by andrross** [comment]
> ❌ Gradle check result for [09778ad](https://github.com/opensearch-project/OpenSearch/commit/09778ad3e563fffeb6a0d76677fbbf93ec5762c1): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/33837/)


https://github.com/opensearch-project/OpenSearch/issues/9208
https://github.com/opensearch-project/OpenSearch/issues/9191

**Review by kkmr — APPROVED** [review]

**Pull request merged**